### PR TITLE
Move permanent part of fkutils here

### PR DESCRIPTION
**Pull request by felixhekhorn**
This PR should have really started at https://github.com/N3PDF/pineko/commit/6d937476c993b29d35188c0f44aedc5be37bafad

- [x] log time taken for each *expensive* operation

### CODE REVIEW

**Review by AleCandido — CHANGES_REQUESTED** [review]
Improve CLI, please :)

File: src/pineko/cli/theory_ekos.py
Comment (felixhekhorn) [reply src/pineko/cli/theory_ekos.py:13]: I can regroup them - but the scope will be theory (as Emmanuele uses "theory")

File: src/pineko/configs.py
Comment (felixhekhorn) [reply src/pineko/configs.py:107]: pylint suggested to do so (R1722) - and it seems ok: https://www.geeksforgeeks.org/python-exit-commands-quit-exit-sys-exit-and-os-_exit/

File: src/pineko/cli/theory_ekos.py
Comment (felixhekhorn) [reply src/pineko/cli/theory_ekos.py]: - things related to theory are now bound below `theory`
- the name `theory` does not relate to the fact, that we're dealing with "theory" stuff here, as opposed to "experimental" - but to the fact that we're expecting a `theory_id` (and hence `theory` is not that bad of a name, although, maybe `TheoryFileAssets` whould be better, but it's too long) and we're iterating over datasets
- now if you say "all we do is like this" - I still want to distinguish between file based operations (`check,compare,convolute,opcard`) and theory based operations; if you really still want to regroup, I can see two options: or the file based move into their own scope and theory becomes root (reverse to the current layout) or we could provide two commands/executables: `pineko` (really meaning `pineko_theory`) and `pineko_file`

**Comment by felixhekhorn** [comment]
I wonder whether to change here:
https://github.com/N3PDF/pineko/blob/784dee8e4a016fc033461fe9ac4e21d1ead67294/src/pineko/configs.py#L79-L83

and add a `else` after the if, because otherwise you get something like this:
```
$ pineko -c /home/a.toml theory opcards 213 LHCB_Z_13TEV_DIELECTRON
Configurations loaded from '/home/a.toml'
Analyze LHCB_Z_13TEV_DIELECTRON
Success: Wrote card with 1 Q2 points to 
/home/felix/Physik/N3PDF/FK/pineko/data/operator_cards/LHCB_DY_13TEV_DIELECTRON.yaml
```
where `/home/a.toml` never existed and the config truely used the default one

**Comment by AleCandido** [comment]
Don't change `detect`: if you don't like this behavior, you should change `load`

https://github.com/N3PDF/pineko/blob/784dee8e4a016fc033461fe9ac4e21d1ead67294/src/pineko/configs.py#L101-L102

or even change where it's used:

https://github.com/N3PDF/pineko/blob/784dee8e4a016fc033461fe9ac4e21d1ead67294/src/pineko/cli/_base.py#L20-L24

Indeed, `detect` is already passing all the information: since it no one existed, it's even raising an error, that is later caught by `load`. So, `detect` has done its job.

I'd suggest not to call `detect` directly inside `load`, but to pass the `detect` result as `load` argument.
Consequently, the exception management should not be done inside `load`, but inside `command`, and so you can decide to print a different message, or even to just fail, and keep propagating the exception.

**Comment by AleCandido** [comment]
> * what is `add_scope` supposed to do?

It's a function I'm using in `runcardsrunner` corresponding instance, to add defaults for an entire section, but preserving values that have explicitly been specified.

> (which would correspond to the `else` I mentioned)

This is false: the `else` would correspond to the case in which no path is specified. In that case either you return an error or you adopt some defaults. What is happening there is that you attempt a few sensible defaults, and if even those are not working, then raise.

> * first, it does not need to be a file, also a path would go through

What's the difference between a path and a file? On CLI you can only write strings, i.e. paths...

> * I dropped the `return {"paths": {"root": pathlib.Path.cwd()}}` because it would result in a ValueError right after as I have some mandatory configs

This is fine, I've defaults for everything in `runcardsrunner`, but it's more of a choice.

> * (finally I also added more docs)

Always a good idea.

File: src/pineko/cli/theory_.py
Comment (AleCandido) [inline src/pineko/cli/theory_.py:35]: Here and below: if you put directly `--no-logs` follow the pattern on the [docs](https://click.palletsprojects.com/en/8.0.x/options/#boolean-flags) and provide both, and set the default to `False`.

File: src/pineko/cli/theory_.py
Comment (AleCandido) [inline src/pineko/cli/theory_.py:39]:
```suggestion
    theory.TheoryBuilder(theory_id, datasets, overwrite).ekos(not logs)
```

File: src/pineko/theory.py
Comment (AleCandido) [inline src/pineko/theory.py:261]:
```suggestion
    def ekos(self, silence):
```

File: src/pineko/cli/theory_.py
Comment (felixhekhorn) [reply src/pineko/cli/theory_.py:35]: `True` you mean? because you said logs by default

File: src/pineko/theory.py
Comment (felixhekhorn) [reply src/pineko/theory.py]: silent it is

**Comment by felixhekhorn** [comment]
@AleCandido @scarlehoff @cschwan @andreab1997 

Dear all,

- I now believe to have all elementary stuff needed to build FK tables moved from https://github.com/NNPDF/fktables/blob/main/fkutils/builder.py to here, please take a look if you can and provide me with some feedback
- a tentative "manual" is given in README.md
- some improvements (@andreab1997 ) over the old implementation:
   - logs are now generated for every run (no more piping needed `python build.py > logs/eko/bla.log`)
   - grids can be inherited by command (instead of manual linking)
   - ekos can be inherited
   - no more prefixing with theory ID (so @scarlehoff you will be able to just copy the folder as is)
   - no more dependency on apfelcomb database but only ymldb (@scarlehoff I'm only reading the members for the rest I don't care)
   - no more special treatment of APPL thanks to ymldb
- some problems still to solve (can be done here or in a seperate PR):
  - allow scale variations
  - add argument to `parser.get_yaml_information` to allow only to ask for the name, so we can run eko without the actual grids being present
  - the hacks I needed for jets (i.e. https://github.com/NNPDF/fktables/blob/main/split_grid.py https://github.com/NNPDF/fktables/blob/main/jets_build.sh https://github.com/NNPDF/fktables/blob/86927f9460f8bbaf7f4daab0789c2d32283d3f2d/fkutils/builder.py#L76 https://github.com/NNPDF/fktables/blob/86927f9460f8bbaf7f4daab0789c2d32283d3f2d/fkutils/builder.py#L173 ) I did *not* port, but I hope https://github.com/N3PDF/eko/pull/105 will be available soon
  - I'd be happy to have some more opinions on https://github.com/N3PDF/pineko/pull/12#discussion_r818837993
- once we're happy there is a final step, i.e. moving the files from the old infrastructure to the new one - which mainly consists in stripping the prefixed theory ID and then to adjust the compare script to the new layout ...

**Comment by cschwan** [comment]
@felixhekhorn cheers, I already had a quick look at the README, see my comments and edits there.

**Comment by felixhekhorn** [comment]
@enocera of course I should have tagged you as well - apologies ;-) please feel free to take a look

File: src/pineko/cli/theory_ekos.py
Comment (AleCandido) [reply src/pineko/cli/theory_ekos.py]: However, I'm perfectly satisfied with the current layout.

The only difference between this and what I suggested is one name: `theory` in place of `compute`. I still think that they are grouped not because they are related to a theory ID, but because they are part of the theory computation process, instead of being a utility command (so called "file based operation").

> * the name `theory` does not relate to the fact, that we're dealing with "theory" stuff here, as opposed to "experimental" - but to the fact that we're expecting a `theory_id`

That's perfect, so `theory_id` is a wonderful name for a positional argument.

File: src/pineko/parser.py
Comment (AleCandido) [inline src/pineko/parser.py:3]: Do we actually need this file at the end? Because I'd like to have a single copy, either here or in `vp`

**Review by AleCandido — COMMENTED** [review]
I still have to dig `pineko.evolve` and `pineko.theory`, but everything else it's good enough.

I left a few comments, mainly about code I suppose to be not needed, and I suggest to just drop.

File: .pylintrc
Comment (felixhekhorn) [reply .pylintrc]: done in [d2018a4](https://github.com/N3PDF/pineko/pull/12/commits/d2018a467c3860212a2636ea71abb7c037290eac)

File: _run.py
Comment (felixhekhorn) [reply _run.py]: done in [24674cf](https://github.com/N3PDF/pineko/pull/12/commits/24674cf9564ecde93532c5c3f18b128446f8de01)

File: data/grids/load.sh
Comment (felixhekhorn) [reply data/grids/load.sh:8]: yes, this is required for debug reasons (as we discussed to download local stuff) - and will be also needed in the future ...

File: src/pineko/__main__.py
Comment (felixhekhorn) [reply src/pineko/__main__.py]: Agreed, it is superseded by the CLI. Dropped in [b69353e](https://github.com/N3PDF/pineko/pull/12/commits/b69353e7296efa6d7673dadbf519fb7b670848f9)

File: src/pineko/parser.py
Comment (scarlehoff) [reply src/pineko/parser.py]: Since we will depend on `eko` and `pineappl` I think it is only natural to depend on `pineko`.

I'm fine with adding extra flags that change the behavior as long as the "default" is what vp is doing. Once this PR is merged and we can test that vp is able to simply import this from pineko and that's it we can remove it from vp.

Comment (scarlehoff) [reply src/pineko/parser.py]: On a second thought, this might mean moving `pineko` to `NNPDF` since it will need to be accessible by everyone (as this will be the package that is able to build and read the theories)

Comment (felixhekhorn) [reply src/pineko/parser.py]: but the package is already public, so everybody can use it ... and maybe sooner or later we want to publish even on PyPI

Comment (AleCandido) [reply src/pineko/parser.py]: No strong opinion on this topic, I'm just glad we all agree to remove duplication

**Review by AleCandido — APPROVED** [review]

**Comment by scarlehoff** [comment]
Once this is merged and a package created I can try and use it in vp.

**Comment by scarlehoff** [comment]
I'm going to merge this if nobody complains.

**Comment by felixhekhorn** [comment]
I opened #13 and #14 for the remaining problems (SV is tracked in #6)

**Comment by scarlehoff** [comment]
Then I can merge, right? As soon as we merge we can tackle the different issues one by one.

**Comment by felixhekhorn** [comment]
For future reference: a bin-by-bin-evolution (as needed for jets) is still possible if needed after https://github.com/N3PDF/eko/pull/105 , but this time I would rather hack outside `pineko` then inside: we could define a bunch of fake-datasets: `CMS_1JET_8TEV-bin00000` containing exactly one member which itself contains exactly one bin /cc @cschwan

**Pull request merged**
